Upgrade script for patching broker annotation

vaikas opened the pull request:
Fixes #2840

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- Introduce a script that runs as a job that will patch all Brokers without BrokerClass annotation to ChannelBasedBroker
-
-

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- 🎁 Add new feature
- 🐛 Fix bug
- 🧽 Update or clean up current behavior
- 🗑️ Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
🗒️ If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
- 🐛 Fix bug
* action required * You have to run the provided upgrade script, or something like this to update all the Brokers with the proper BrokerClass or they will not continue to be reconciled after 0.14 cuts.
Details for the script are in:
config/upgrade/v0.14.0/README.md
Release artifact is:
https://github.com/knative/eventing/releases
```

**Docs**

<!--
📖 If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

mattmoor commented on config/upgrade/v0.14.0/README.md:
```diff
@@ -0,0 +1,29 @@
+# Upgrade script required to upgrade to v0.14.0 of Eventing
+
+This directory contains a job that upgrades (patches) all
+the Brokers to have an annotation that specifies which
+BrokerClass should be reconciling it. Prior to v0.13.0 there
+was only the ChannelBasedBroker and it was not necessary.
+However in v0.14.0 we introduced BrokerClass and in order
+for existing Brokers to continue to be reconciled, they
+need to be patched to include the BrokerClass annotation
+like so.
```
<!--mdfmt-bot-->
Format markdown:
```suggestion
This directory contains a job that upgrades (patches) all the Brokers to have an
annotation that specifies which BrokerClass should be reconciling it. Prior to
v0.13.0 there was only the ChannelBasedBroker and it was not necessary. However
in v0.14.0 we introduced BrokerClass and in order for existing Brokers to
continue to be reconciled, they need to be patched to include the BrokerClass
annotation like so.
```

mattmoor replied:
<!--mdfmt-bot-->
Format markdown:
```suggestion
This directory contains a job that upgrades (patches) all the Brokers to have an
annotation that specifies which BrokerClass should be reconciling it. Prior to
v0.13.0 there was only the ChannelBasedBroker and it was not necessary. However
in v0.14.0 we introduced BrokerClass and in order for existing Brokers to
continue to be reconciled, they need to be patched to include the BrokerClass
annotation like so.
```

mattmoor replied:
<!--mdfmt-bot-->
Format markdown:
```suggestion
This directory contains a job that upgrades (patches) all the Brokers to have an
annotation that specifies which BrokerClass should be reconciling it. Prior to
v0.13.0 there was only the ChannelBasedBroker and it was not necessary. However
in v0.14.0 we introduced BrokerClass and in order for existing Brokers to
continue to be reconciled, they need to be patched to include the BrokerClass
annotation like so.
```

mattmoor commented on config/upgrade/v0.14.0/README.md:
````diff
@@ -0,0 +1,29 @@
+ko apply -f ./upgrade.yaml
+```
+
+It will create a job called v0.14.0-upgrade in the knative-eventing
+namespace. If you installed to different namespace, you need to
+modify the upgrade.yaml appropriately. Also the job by default
+runs as `eventing-controller` service account, you can also modify
+that but the service account will need to have permissions
+to list `Namespace`s, list and patch `Broker`s.
````
<!--mdfmt-bot-->
Format markdown:
```suggestion
It will create a job called v0.14.0-upgrade in the knative-eventing namespace.
If you installed to different namespace, you need to modify the upgrade.yaml
appropriately. Also the job by default runs as `eventing-controller` service
account, you can also modify that but the service account will need to have
permissions to list `Namespace`s, list and patch `Broker`s.
```

mattmoor replied:
<!--mdfmt-bot-->
Format markdown:
```suggestion
It will create a job called v0.14.0-upgrade in the knative-eventing namespace.
If you installed to different namespace, you need to modify the upgrade.yaml
appropriately. Also the job by default runs as `eventing-controller` service
account, you can also modify that but the service account will need to have
permissions to list `Namespace`s, list and patch `Broker`s.
```

mattmoor replied:
<!--mdfmt-bot-->
Format markdown:
```suggestion
It will create a job called v0.14.0-upgrade in the knative-eventing namespace.
If you installed to different namespace, you need to modify the upgrade.yaml
appropriately. Also the job by default runs as `eventing-controller` service
account, you can also modify that but the service account will need to have
permissions to list `Namespace`s, list and patch `Broker`s.
```

grantr commented:
/cc @liu-cong

mattmoor commented on config/upgrade/v0.14.0/README.md:
```diff
@@ -0,0 +1,26 @@
+If you installed to different namespace, you need to modify the upgrade.yaml
+appropriately. Also the job by default runs as `eventing-controller` service
+account, you can also modify that but the service account will need to have
+permissions to list `Namespace`s, list and patch `Broker`s. 
```
<!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
permissions to list `Namespace`s, list and patch `Broker`s.
```

mattmoor replied:
<!--mdfmt-bot-->
Format markdown:
```suggestion
permissions to list `Namespace`s, list and patch `Broker`s.
```

matzew commented on config/upgrade/v0.14.0/README.md:
````diff
@@ -0,0 +1,26 @@
+annotation like so.
+
+```
+  annotations:
````
@k4leung4  I think this is someting that the operator should do
/cc @aliok

matzew commented:
will this be removed _after_ the release ?

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
```
nit: remove empty line

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+
+// Upgrade upgrades all the brokers by applying an Annotation to all the
+// ones that do not have them. This is necessary to ensure that existing Brokers
+// that do not have Annotation will continue to be reconciled by the existing
```
nit: ...that do  not have the `eventing.knative.dev/broker.class` annotation

liu-cong commented on cmd/upgrade/v0.14.0/main.go:
```diff
@@ -0,0 +1,41 @@
+	cfg := sharedmain.ParseAndGetConfigOrDie()
+	ctx = context.WithValue(ctx, kubeclient.Key{}, kubernetes.NewForConfigOrDie(cfg))
+	ctx = context.WithValue(ctx, eventingclient.Key{}, versioned.NewForConfigOrDie(cfg))
+	broker.Upgrade(ctx)
```
handle error?

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+		logger.Infof("Patching Broker \"%s/%s\" with %q", broker.Namespace, broker.Name, string(patch))
+		patched, err := brokerClient.Patch(broker.Name, types.MergePatchType, patch)
+		if err != nil {
+			logger.Warnf("Failed to patch \"%s/%s\" : %v", broker.Namespace, broker.Name, err)
```
return error here?

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+		logger.Warnf("Failed to create patch for \"%s/%s\" : %v", broker.Namespace, broker.Name, err)
+		return []byte{}, err
+	}
+	logger.Infof("Patch: %q", string(patch))
```
nit: also log the broker name and namespace

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+	logger.Infof("Patch: %q", string(patch))
+	// If there is nothing to patch, we are good, just return.
+	// Empty patch is {}, hence we check for that.
+	if len(patch) <= 2 {
```
Can this happen in practice?

liu-cong replied:
I mean in this code looks like we can only come to here if the broker is missing the annotation, so looks like the patch will always be non-empty.

grantr replied:
You're right, it doesn't look like this will ever happen. We could maybe confirm that by looking at the coverage report to see if this line is covered.

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+	// No annotations, or missing, add it...
+	modified := broker.DeepCopy()
+	modifiedAnnotations := modified.ObjectMeta.GetAnnotations()
+	if modifiedAnnotations == nil {
+		modifiedAnnotations = make(map[string]string, 1)
+	}
+	if _, present := modifiedAnnotations[eventing.BrokerClassKey]; !present {
```
nit: code structure can be simplified if you do `modified := broker.DeepCopy()` at the beginning so you only need to check the modified copy. You waste a copy if nothing needs to be patched but you can avoid checking the annotation twice.

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader_test.go:
```diff
@@ -0,0 +1,203 @@
+	"context"
+	"reflect"
+
+	fakeeventingclient "knative.dev/eventing/pkg/client/injection/client/fake"
```
nit: this one can be moved down?

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader_test.go:
```diff
@@ -0,0 +1,203 @@
+		expectPatch []byte
+	}{
+		"nilannotations": {
+			&v1alpha1.Broker{ObjectMeta: metav1.ObjectMeta{Name: testbroker}},
```
nit: use `broker(testbroker, testns, nil)` for consistency

vaikas replied:
I wanted it without annotations to test the case where annotations is nil. broker( sets the annotation).

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader_test.go:
```diff
@@ -0,0 +1,203 @@
+	}{
+		"nilannotations": {
+			&v1alpha1.Broker{ObjectMeta: metav1.ObjectMeta{Name: testbroker}},
+			[]byte("{\"metadata\":{\"annotations\":{\"eventing.knative.dev/broker.class\":\"ChannelBasedBroker\"}}}"),
```
nit: `patchbytes` already defined so just use it

liu-cong commented on pkg/upgrader/broker/v0.14.0/upgrader.go:
```diff
@@ -0,0 +1,124 @@
+	return nil
+}
+
+func ProcessNamespace(ctx context.Context, ns string) error {
```
nit: this probably doesn't need to be exported.

grantr commented:
@vaikas do we need a docs PR too? Since the install docs link directly to release artifacts, anyone following the install instructions to upgrade won't see that this is required. To be fair that seems like a general docs problem and not specific to this issue.

vaikas commented:
I don't see a need to remove it after the release, that's why I versioned it. But thoughts?

vaikas commented:
@liu-cong thanks, I believe I addressed all comments, PTAL.

@grantr Seems like we should have a separate "upgrade" section and I don't see a place for it, seems like in the past the ReleaseNotes has been a place where we've tacked things that require attention and from what I understand that's been working?

liu-cong commented:
/lgtm
